Sources_v3/StreamChat/ChatClient.swift

```diff
@@ -219,15 +210,26 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
     
     /// An array of requests waiting for the connection id
     @Atomic var connectionIdWaiters: [(String?) -> Void] = []
+
+    /// An array of requests waiting for the token
+    @Atomic var tokenWaiters: [(Token?) -> Void] = []
     
     /// The token of the current user. If the current user is anonymous, the token is `nil`.
     @Atomic var currentToken: Token?
+
+    public var tokenProvider: _TokenProvider<ExtraData>
     
     /// Creates a new instance of `ChatClient`.
-    ///
-    /// - Parameter config: The config object for the `Client`. See `ChatClientConfig` for all configuration options.
-    ///
-    public convenience init(config: ChatClientConfig) {
+    /// - Parameters:
+    ///   - config: The config object for the `Client`. See `ChatClientConfig` for all configuration options.
+    ///   - tokenProvider: The `_TokenProvider<ExtraData>` instance that incapsulates the logic of obtaining a JWT
+    ///   token used to communicate with REST-API.
+    ///   - didReloadUser: The closure that will be called once the **first** session for the token is obtained and user session is setup.
+    public convenience init(
+        config: ChatClientConfig,
+        tokenProvider: _TokenProvider<ExtraData>,
+        didReloadUser: @escaping (Error?) -> Void = { _ in }
```

Comment:
Nice idea. Can we maybe rename it just to `completion` and also make it optional? We will call it once we're done with set up.

Reply:
On the other hand - check my Slack comments. This will need to be benchmarked.